File: Golang/Concurrency/Concurrency_basics.go
package main

import (
	"fmt"
	"sync"
)

func pradeep(wg *sync.WaitGroup, name string) {
	defer wg.Done()
	for {
		fmt.Println("Pradeep is continously working")
		fmt.Printf("Pradeep Running this in a new Thread: %s\n", name)
	}
}

func main() {
	var wg sync.WaitGroup
	wg.Add(2)
	go pradeep(&wg, "goroutine-1")
	go pradeep(&wg, "goroutine-2")

	// Alternative approaches considered:
	//
	// A fixed-size worker pool running a single task:
	//   task1 := func() {
	//       for i := 0; i < 20; i++ {
	//           fmt.Println("pradeep")
	//       }
	//   }
	//
	// Three tasks each run on their own goroutine, then joined:
	//   task1 := func() { for i := 0; i < 20; i++ { fmt.Println("pradeep") } }
	//   task2 := func() { for i := 0; i < 20; i++ { fmt.Println("ishita") } }
	//   task3 := func() { for i := 0; i < 20; i++ { fmt.Println("Sahil") } }
	//   var innerWg sync.WaitGroup
	//   for _, task := range []func(){task1, task2, task3} {
	//       innerWg.Add(1)
	//       go func(t func()) { defer innerWg.Done(); t() }(task)
	//   }
	//   innerWg.Wait()
	//   fmt.Println("main thread")

	wg.Wait()
}

File: JavaScript/Concurrency/Concurrency_basics.js
'use strict';

const { Worker, isMainThread, workerData } = require('worker_threads');

function runWorker() {
    while (true) {
        console.log('Pradeep is continously working');
        console.log(`Pradeep Running this in a new Thread: ${workerData.name}`);
    }
}

function main() {
    // JS has no way to run two truly infinite, blocking loops in parallel on
    // the main thread (it's single-threaded), so real OS threads are used
    // here to reproduce Java's two concurrently-running Thread objects.
    new Worker(__filename, { workerData: { name: 'worker-1' } });
    new Worker(__filename, { workerData: { name: 'worker-2' } });

    // Alternative approaches considered:
    //
    // A fixed-size worker pool running a single task:
    //   const task1 = () => { for (let i = 0; i < 20; i++) console.log('pradeep'); };
    //
    // Three tasks each run on their own worker, then joined:
    //   const task1 = () => { for (let i = 0; i < 20; i++) console.log('pradeep'); };
    //   const task2 = () => { for (let i = 0; i < 20; i++) console.log('ishita'); };
    //   const task3 = () => { for (let i = 0; i < 20; i++) console.log('Sahil'); };
    //   await Promise.all([task1, task2, task3].map((t) => new Promise((resolve) => {
    //       const w = new Worker(...);
    //       w.on('exit', resolve);
    //   })));
    //   console.log('main thread');
}

if (isMainThread) {
    main();
} else {
    runWorker();
}
